### PR TITLE
Implement selections in CLineInput

### PR DESCRIPTION
**Pull request by Learath2**
Baby steps towards #12, leaving a draft for feedback.

### CODE REVIEW

File: src/game/client/lineinput.cpp [hunk 1]
```diff
@@ -177,6 +188,18 @@ bool CLineInput::Manipulate(IInput::CEvent Event, char *pStr, int StrMaxSize, in
 				}
 			}
 		}
+		else if((pInput->KeyIsPressed(KEY_LCTRL) || pInput->KeyIsPressed(KEY_RCTRL)) && Key == KEY_C)
+		{
+			char aBuf[MAX_SIZE];
+			int Len = CursorPos - SelectionStart;
+			str_format(aBuf, sizeof(aBuf), "%.*s", Len > 0 ? Len : Len * -1, &pStr[Len > 0 ? SelectionStart : CursorPos]);
```
Comment (Robyt3) [inline src/game/client/lineinput.cpp]: `Len > 0 ? Len : Len * -1` could be replaced with `abs(Len)`

File: src/game/client/lineinput.cpp [hunk 2]
```diff
@@ -37,6 +38,13 @@ void CLineInput::Set(const char *pString)
 	}
 }
 
+void CLineInput::GetSelection(char *pBuf, int BufSize) const
+{
+	int Len = m_CursorPos - m_SelectionStart;
+
+	str_format(pBuf, BufSize, "%.*s", Len > 0 ? Len : Len * -1, &m_Str[Len > 0 ? m_SelectionStart : m_CursorPos]);
```
Comment (Robyt3) [inline src/game/client/lineinput.cpp]: `absolute(Len)`

**Comment by Robyt3** [comment]
When you have an active selection and then insert text without holding shift, it should override the selection with the new char. Also when having a selection and pressing delete/backspace.
There is also a bug when entering a string, holding shift and then pressing backspace, it selects the empty characters.

**Comment by Learath2** [comment]
Thanks @Robyt3

**Comment by Robyt3** [comment]
Good work! Some more minor issues:
- Any text navigation, when shift isn't pressed, should clear the selection, even if this doesn't change the selected position. For instance, having everything selected and the cursor at the end, if you press right arrow (without shift pressed) it should also clear the selection.
- Pasting a text over a selection is glitchy if the text to paste doesn't fit in the max length. Take the quick search for example, enter the maximum number of characters randomly. Select like 5 characters, copy them to the clipboard, then select only one character and paste over it. That replaces the selected character, but the total length of the string seems to decrease and the selection is weirdly retained, which allows you to paste over the same inserted character again.
- Ctrl+A and Ctrl+X would also be nice.

File: src/test/str.cpp [hunk 3]
```diff
@@ -124,3 +124,15 @@ TEST(Str, PathUnsafe)
...
+TEST(Utf8, CharacterCount)
+{
+	EXPECT_EQ(str_utf8_charcount("", 0), 0);
+	EXPECT_EQ(str_utf8_charcount("", 1), 0);
+	EXPECT_EQ(str_utf8_charcount("a", 1), 1);
+	EXPECT_EQ(str_utf8_charcount("a", 2), 1);
+	EXPECT_EQ(str_utf8_charcount("a", 3), 1);
+	EXPECT_EQ(str_utf8_charcount("ab", 2), 2);
+	EXPECT_EQ(str_utf8_charcount("abc", 2), 2);
+	EXPECT_EQ(str_utf8_charcount("\xE2\x86\x91\xF0\x9F\x98\x81", 7), 2);
```
Comment (heinrich5991) [inline src/test/str.cpp]: You could put the raw characters here, compilers can handle UTF-8. ;)

**Comment by Learath2** [comment]
I'm really not very proud with the logic here but I didn't really come up with a good way to clean it up, lmk if anyone sees a way.

**Comment by Robyt3** [comment]
- Pressing Home/End should also deselect, like left/right arrows do, even when the cursor wouldn't move as a result.
- Crashes when overwriting with empty string. Enter any characters. Press Ctrl+C (without selection), to copy the empty string to the clipboard. Sometimes the selection is retained when pressing Ctrl+V, which results in null chars being selected.  Pressing Ctrl+V again with the invalid selection crashes.
  - Ctrl+C should not copy the empty string. Normal UX is to not copy anything if nothing is selected.
  - Ctrl+V should properly adjust the selection length. The bug only seems to happen when overwriting characters in the second half of the entered string, especially when overwriting the last character.

**Comment by Robyt3** [comment]
LGTM now!

**Comment by oy** [comment]
- When there's a line break in the console the cursor isn't shown anymore and when you select sth it's not rendered correctly.
- When the line breaks in chat the selection isn't rendered correctly. If you press the down arrow afterwards to erase the chat, write something, then select what you wrote and press another left the text is rendered twice.

**Comment by Learath2** [comment]
@oy don't know how to reproduce the console one as I can't get linebreaks in the console at all. I couldn't reproduce the text rendering twice either but the last commit might have fixed that.

**Comment by oy** [comment]
@Learath2 try a non-widescreen resolution like 1024x768 and write lots of wide characters like `w`.
Will check out the changes later.

**Comment by Learath2** [comment]
@oy I think this finishes it up, though I'm not sure if you still want to get this in with new textrender improvements on the way that could make it much cleaner.

**Comment by oy** [comment]
@Learath2 yeah think we should wait till the new textrenderer is merged.

**Comment by Learath2** [comment]
@oy you should still cherry-pick e927f00, that's independent of the selection

**Comment by Learath2** [comment]
#2825 supersedes this

**Pull request closed without merging**